### ISSUE TITLE
Prevent autoraising while dragging or resizing a window

### ISSUE THREAD

**Issue by dreeves** [sbmpost/AutoRaise#8]
Bug report!

## Reseminata (steps to reproduce)

1. Take an iterm2 window that overlaps a google chrome window (I don't think it matters which apps)
2. Click and hold the top edge of the iterm2 window at a spot where it's in front of the chrome window
3. Slowly drag up and down to make the iterm2 window taller and shorter

## Expectata (what I expected)

I'm always on top of (the edge of) the iterm2 window so it should keep focus.

## Resultata (what resulted instead)

The chrome window jumps in front of the iterm2 window!

**Comment by sbmpost** [comment]
Yes, I can reproduce this. Not sure if I can fix it though ;-) Let me see what I can do. It is going to take some time though as I am very busy ATM.

**Comment by sbmpost** [comment]
This should be fixed now, you can download the latest binary here:
https://github.com/sbmpost/AutoRaise/raw/master/AutoRaise

**Comment by dreeves** [comment]
Fix confirmed!

**Comment by sbmpost** [comment]
Cool :-) closing this issue then and thanks for reporting!

**Comment by sbmpost** [comment]
@dreeves I have been playing around with some experimental code that prevents a smaller window from losing the focus if it is fully contained within a larger window:

https://github.com/sbmpost/AutoRaise/pull/10.

For example, you are working in a large "TextEdit" window, and you activate a smaller "Notes" window that was hiding below. In such case you will be able to reach the "Notes" window without the TextEdit window being raised again. I am curious what you think if this small tweak and if I should include it in the master branch.

**Comment by dreeves** [comment]
In theory it sounds like the kind of special exception that can be a bad idea. But, in practice, it sounds quite smart! I'm game to try it if you want to push it to master!

**Comment by sbmpost** [comment]
@dreeves you can find the binary here:
https://github.com/sbmpost/AutoRaise/raw/3833106518d967ae604ee73cd34464fd6fa944bf/AutoRaise
Thanks for willing to try it out!

**Comment by dreeves** [comment]
It's working great! I say push it to master!

(Not part of this gissue but it's still frustrating that autoraising happens sometimes when you don't want it to. Like when doing something at the edge of window. I mean due to imprecise mousing, not a bug. I may increase AutoRaise's delay but that's also not great. As we talked about elsewhere, ideally the focus-follows-mouse part would happen immediately with the autoraise part happening with a longer (like 500ms) delay!)

**Comment by sbmpost** [comment]
@dreeves 
> It's working great! I say push it to master!
Cool :-) nice to hear!

> Like when doing something at the edge of window.
I added some more experimental code which results in an artificial border of 10 pixels around the currently focussed window. Only when the mouse hovers outside of this border, the window below it will be raised. I am not sure if 10 pixels is a good value, and it also doesn't cover all imprecise mousing cases. But it might improve the situation somewhat. You can download again from here:

https://github.com/sbmpost/AutoRaise/blob/3833106518d967ae604ee73cd34464fd6fa944bf/AutoRaise

> As we talked about elsewhere, ideally the focus-follows-mouse part would happen immediately with the autoraise part happening with a longer (like 500ms) delay!)
Indeed, but unfortunately according to Apple, focussing and raising are the same thing.

**Comment by dreeves** [comment]
> an artificial border of 10 pixels around the currently focussed window

I'm trying it out and I think it would be safe to add a fair bit more padding!

> according to Apple, focussing and raising are the same thing

Boo Apple! :(

**Comment by sbmpost** [comment]
@dreeves I have doubled the artificial border size from 10 to 20 pixels.
Latest version is here:
https://github.com/sbmpost/AutoRaise/blob/contained-window-stays-on-top/AutoRaise

**Comment by dreeves** [comment]
Seems like an improvement! 

PS: Not sure if the following is worth looking into -- possibly it's even desired behavior -- but:

## Reseminata

1. Put one window flush with the right edge of the screen (call it window BFR for "background flush right")
2. Put another window on top of it within a few pixels of the right edge (call it CWP for "close, within pixels")
3. Move the mouse to the right edge of the screen

## Expectata

The mouse is within a few pixels of CWP, the currently focused window, and so window BFR should not autoraise. I should still be able to click to raise it.

## Resultata

Window BFR autoraises.

**Comment by sbmpost** [comment]
@dreeves it took me some time to find out what you meant, but yes I can reproduce this. I will be investigating this further to see if it can/should be fixed. I am also considering another change. Currently there are two rules in effect:

1. The artificial border of 20 px when the mouse is about to leave the focused window
2. Don't raise the background window if the focused window is fully contained within it

To keep things simple, I am considering to replace these rules with this one:

Don't raise the background window if the focused window is fully contained within it, but before doing the check, first add a 20 px artificial border to the background window. This would mean however that imprecise mousing is only dealt with when the focussed window is contained within the background window. So far this seems to work quite well. I will push a new version somewhere in the coming days.

Thanks for all the feedback you are giving me, it is really helpful :-)

**Comment by sbmpost** [comment]
Actually after writing the previous message, I realised that my proposed change, will also mitigate (not completely solve) the situation you were describing: The moment you put the background window flush with the right edge of the screen, it will no longer be raised as long as the focused window falls of the right edge but is still within the 20 px range. So I am pushing this version now for you to try out:

https://github.com/sbmpost/AutoRaise/blob/contained-window-stays-on-top/AutoRaise

**Comment by dreeves** [comment]
Ah, thank you! It's great to feel like I'm contributing in a small way to these steady improvements! Really brilliant work here.

I do have a quibble with the latest version. I like to have windows peeking out behind other windows, like so:

![image](https://user-images.githubusercontent.com/9928/101992176-24671c00-3c66-11eb-822a-b4cd783a018b.png)

That way I can, in this example, shove my mouse to the top of the screen and my editor jumps in front of my browser window. Which is great. But in this new version of AutoRaise, that editor windows is treated (if I've understood correctly) as fully contained in the browser window, despite peeking out, so it doesn't drop back behind the browser window unless I click on the browser window. (Would it help for me to file a separate gissue for that?)

I don't know if it's possible to have the best of all worlds but I think for my own use case I'd rather sacrifice the "suppress autoraise if current window would become totally obscured" feature.

**Comment by sbmpost** [comment]
> that editor windows is treated (if I've understood correctly) as fully contained in the browser window, despite peeking out, so it doesn't drop back behind the browser window unless I click on the browser window.

Well, the 20px condition must be met on all sides of the window in order to suppress the raising. From what I can tell in the example screenshot, this condition is not met. So the browser window will be raised over the editor window because the editor window is more than 20 pixels away from it on the left side. Or did you perhaps deliberately add the extra space on the left to clarify what is going on in the screenshot? I can see your point however and maybe the artificial border should not be considered for the "totally obscured" check. In fact, I think what you are saying is that the previous version was working slightly better for your use case correct?

**Comment by sbmpost** [comment]
@dreeves ok, I looked again at your screenshot, I wasn't looking sharply :-) I do have an idea however. What if the contained with check would only consider the left and right (and maybe also the bottom) sides? In addition to that the 20 px artificial border for the case of putting a background window flush with the right edge of the screen? Let me make those changes...

**Comment by dreeves** [comment]
Having AutoRaise treat top/bottom/left/right differently sounds wrong to me. (Also adding too much confusion and complexity even if it ended up satisfying my use cases!)

Proposal:

1. The artificial border is used to prevent the current window from losing focus if the mouse is only barely outside it. (If I'm barely outside a window, that's not intentionally leaving the window, that's just just imprecise mousing.)
2. Exception: if the mouse is at the edge of the screen, that's not imprecise mousing, that's intentional. (Conceptually, having the mouse at the screen edge is like having it way past the screen edge. That's where I'm trying to go and am only stopped by the screen edge.)
3. The "fully contained" exception only applies if the current window is truly fully contained -- no artificial border. (We want to keep a window from totally disappearing. If it's peeking out, it's not totally disappearing.)

**Comment by sbmpost** [comment]
I tried my idea, and I didn't like the left/right/bottom artificial border either. Sometimes the artificial border is just 'in the way' instead of helpful. With this in mind, let me go over your 3 points.

Point 1: True, but just like you, I keep my background windows only a few pixels above the focused window. If I move my mouse to that small pixel area, I expect the background window to be raised. So I guess this relates to your second point because these small pixel areas are typically near the edge of the screen, and therefore not to be considered for imprecise mousing. As for point 3: I totally agree.

In conclusion: For now I will get rid of the artificial border completely because it only makes sense if you can decide better when to apply it. As you already suggested, we better treat that as a separate issue/improvement.

**Comment by dreeves** [comment]
Ah, phew, I like the latest version much better! Thank you again!

I think imprecise mousing mitigation could improve it further if done right. I feel like I really convinced myself about my proposal above. So, yeah, lemme gissue that separately!

**Comment by sbmpost** [comment]
cool :-) for now I am merging the changes with master. I don't know when I will find time to focus on #14 but at least we have this documented. Thank you!

**Comment by dreeves** [comment]
I think there may have been a regression here. In my use case where a window is peeking out (see [screenshot](https://github.com/sbmpost/AutoRaise/issues/8#issuecomment-743798853)), it no longer autoraises.
